pennylane/templates/subroutines/permute.py

```diff
@@ -0,0 +1,184 @@
+            from pennylane.templates import Permute
+
+            with qml.tape.QuantumTape() as tape:
+                # RZs added to ensure numerical ordering in drawing
+                for wire in range(5):
+                    qml.RZ(0, wires=wire)
+                Permute([4, 2, 0, 1, 3], wires=[0, 1, 2, 3, 4])
+
+
+        >>> print(tape.draw())
+        0: ──RZ(0)─────────╭SWAP────────────────┤
+        1: ──RZ(0)──╭SWAP──│────────────────────┤
+        2: ──RZ(0)──╰SWAP──│──────╭SWAP─────────┤
+        3: ──RZ(0)─────────│──────│──────╭SWAP──┤
+        4: ──RZ(0)─────────╰SWAP──╰SWAP──╰SWAP──┤
+
+        ``Permute`` can also be applied to wires with arbitrary labels, like so:
+
+        .. code-block:: python
+
+            wire_labels = [3, 2, "a", 0, "c"]
+
+            with qml.tape.QuantumTape() as tape:
+                # RZs added to ensure numerical ordering in drawing
+                for wire in range(5):
+                    qml.RZ(0, wires=wire)
+                Permute(["c", 3, "a", 2, 0], wires=wire_labels)
+
+        The permuted circuit is:
+
+        >>> print(tape.draw())
+        3: ──RZ(0)──╭SWAP────────────────┤
+        2: ──RZ(0)──│──────╭SWAP─────────┤
+        a: ──RZ(0)──│──────│─────────────┤
+        0: ──RZ(0)──│──────│──────╭SWAP──┤
+        c: ──RZ(0)──╰SWAP──╰SWAP──╰SWAP──┤
+
+        It is also possible to permute a subset of wires by
+        specifying a subset of labels. For example,
+
+        .. code-block:: python
+
+           wire_labels = [3, 2, "a", 0, "c"]
+
+            with qml.tape.QuantumTape() as tape:
+                # Create 5 wires
+                for wire in range(num_wires):
+                    qml.RZ(0, wires=wire)
+
+                # Only permute the order of 3 of them
+                Permute(["c", 2, 0], wires=[2, 0, "c"])
+
+        will permute only the second, third, and fifth wires as follows:
+
+        >>> print(tape.draw())
+        3: ──RZ(0)────────────────┤
+        2: ──RZ(0)──╭SWAP─────────┤
+        a: ──RZ(0)──│─────────────┤
+        0: ──RZ(0)──│──────╭SWAP──┤
+        c: ──RZ(0)──╰SWAP──╰SWAP──┤
+
+    """
+
+    wires = Wires(wires)
+
+    if len(permutation) <= 1 or len(wires) <= 1:
+        raise ValueError("Permutations must involve at least 2 qubits.")
+
+    # Make sure the length of the things are the same
+    if len(permutation) != len(wires):
+        raise ValueError("Permutation must specify outcome of all wires.")
+
+    # Permutation order must contain all unique values
+    if len(set(permutation)) != len(permutation):
+        raise ValueError("Values in a permutation must all be unique.")
+
+    # Make sure everything in the permutation has an associated label in wires
+    for label in permutation:
+        if label not in wires.labels:
+            raise ValueError(f"Cannot permute wire {label} not present in wire set.")
+
+    # Temporary storage to keep track as we permute
+    working_order = list(wires.labels)
```

Comment:
What happens if a user inputs to the `wires` argument a list of wires that is not the same order as the device's wires?
E.g.,
```python
wires_dev = ["alice", "bob", 5, 9]

@qml.qnode(dev)
def circuit():
    ...
    qml.templates.Permute(["alice", "bob"], wires=["bob", "alice"])
    ...
```
Will it break? I guess there's no way of checking a Device's wires from within a template?

Reply:
It actually seems to work (!), I think because the permutation is done at the level of wire labels and not wire indices. For example,

```
wires_dev = ["alice", "bob", 5, 9]
dev = qml.device('default.qubit', wires=wires_dev)

@qml.qnode(dev)
def circuit():
    for wire in wires_dev:
        qml.RZ(0, wires=wire)
    qml.templates.Permute(["alice", 9, "bob"], wires=[9, "bob", "alice"])
    return qml.expval(qml.PauliZ(wires=5))
```

gives the circuit

```
 alice: ──RZ(0)──╭SWAP──╭SWAP──┤     
   bob: ──RZ(0)──│──────╰SWAP──┤     
     5: ──RZ(0)──│─────────────┤ ⟨Z⟩ 
     9: ──RZ(0)──╰SWAP─────────┤     
```

Even though the original order is ["alice", "bob", 9], the permutation indicates that the contents of the wire 9 should be moved to wherever wire "bob" was, contents of wire "bob" go to where wire "alice" originally was.

Reply:
That's pretty cool!